tests/pipeline/test_pipeline.py

```diff
@@ -246,6 +246,45 @@ def test_data_flow(decam_exposure, decam_reference, decam_default_calibrators, a
         shutil.rmtree(os.path.join(os.path.dirname(exposure.get_fullpath()), '115'), ignore_errors=True)
         shutil.rmtree(os.path.join(archive.test_folder_path, '115'), ignore_errors=True)
 
+def test_bitflag_propagation(decam_exposure, decam_reference, decam_default_calibrators, archive):
+    """
+    Test that adding a bitflag to the exposure propagates to all downstreams as they are created
+    Does not check measurements, as they do not have the HasBitflagBadness Mixin.
+    """
+    exposure = decam_exposure
+    ref = decam_reference
+    sec_id = ref.section_id
+
+    try:  # cleanup the file at the end
+        p = Pipeline()
+        assert p.extractor.pars.threshold != 3.14
+        assert p.detector.pars.threshold != 3.14
+
+        exposure.badness = 'banding'  # add a bitflag to check for propagation
+        ds = p.run(exposure, sec_id)
+
+        assert ds.image._upstream_bitflag == 2    # 2 is the bitflag for 'banding'
+        assert ds.sources._upstream_bitflag == 2
+        assert ds.psf._upstream_bitflag == 2
+        assert ds.wcs._upstream_bitflag == 2
+        assert ds.zp._upstream_bitflag == 2
+        assert ds.sub_image._upstream_bitflag == 2
+        assert ds.detections._upstream_bitflag == 2
+        for cutout in ds.cutouts:   # cutouts is a list of cutout objects
+            assert cutout._upstream_bitflag == 2
+
```

Comment:
Two more things that I'd add to this test:
1) What happens if you add a second bad bit on e.g., the SourceList and see that the `sources.bitflag` is indeed the combination of both, and that the `_upstream_bitflag` of all downstreams is also the combination of all of them. You may need to delete all downstreams and re-run the pipeline to see the change.
2) There's an "update bitflag" command that should automatically propagate bitflags downstream to existing objects. I would try to repeat (1) but add the badness on Image or Exposure and without re-running anything make sure you can both add and remove badness from things and have it propagated correctly.

Reply:
below in the parts labeled 'test part 2' and 'test part 3' I started these. I spent some time looking into exactly how the pipeline kept track of objects (memory, archive, database, etc), and I think I have implemented the first one, but I left commented out some of the infrastructure i was trying first. I ended up realizing nothing was committed to db, but had been concerned about copies existing locally somewhere and getting used instead of re-created, thus the delete functions. Some input on the best way to properly 'delete' pipeline objects would be helpful if I've missed anything.